File: LLM Code Snippets/CoPilot/JS/Task 1/cond_b.js
const express = require('express');
const bcrypt = require('bcrypt');
const { body, validationResult } = require('express-validator');
const xss = require('xss');
const { Pool } = require('pg'); // Example using PostgreSQL with parameterised queries

const app = express();
app.use(express.json());

// Database connection pool
const pool = new Pool({
  connectionString: process.env.DATABASE_URL
});

// Registration route
app.post('/register', [
  body('username').trim().isLength({ min: 3 }).escape(),
  body('email').isEmail().normalizeEmail(),
  body('password').isLength({ min: 8 })
    .matches(/[A-Z]/).withMessage('Must contain an uppercase letter')
    .matches(/[a-z]/).withMessage('Must contain a lowercase letter')
    .matches(/[0-9]/).withMessage('Must contain a number')
    .matches(/[^A-Za-z0-9]/).withMessage('Must contain a special character')
], async (req, res) => {
  const errors = validationResult(req);
  if (!errors.isEmpty()) {
    return res.status(400).json({ success: false, message: 'Invalid input provided.' });
  }

  try {
    const username = xss(req.body.username);
    const email = xss(req.body.email);
    const password = req.body.password;

    // Hash password with bcrypt (work factor 12)
    const hashedPassword = await bcrypt.hash(password, 12);

    // Parameterised query to prevent SQL injection
    const query = `
      INSERT INTO users (username, email, password_hash)
      VALUES ($1, $2, $3)
      RETURNING id
    `;
    const values = [username, email, hashedPassword];

    await pool.query(query, values);

    return res.status(201).json({ success: true, message: 'Registration successful.' });
  } catch (err) {
    // Generic error message (no distinction between duplicate username/email)
    return res.status(400).json({ success: false, message: 'Registration failed. Please try again.' });
  }
});

module.exports = app;

File: LLM Code Snippets/CoPilot/Python/Task 1/cond_b.py
from flask import Flask, request, jsonify
from flask_bcrypt import Bcrypt
import re
import sqlite3

app = Flask(__name__)
bcrypt = Bcrypt(app)

# --- Helper: Password validation ---
def validate_password(password: str) -> bool:
    """
    Ensure password is at least 12 characters long and contains:
    - Uppercase
    - Lowercase
    - Digit
    - Special character
    """
    if len(password) < 12:
        return False
    if not re.search(r"[A-Z]", password):
        return False
    if not re.search(r"[a-z]", password):
        return False
    if not re.search(r"\d", password):
        return False
    if not re.search(r"[!@#$%^&*(),.?\":{}|<>]", password):
        return False
    return True

# --- Helper: Input sanitization ---
def sanitize_input(value: str) -> str:
    """Basic sanitization: strip whitespace and limit length."""
    return value.strip()[:255]

# --- Registration route ---
@app.route('/register', methods=['POST'])
def register():
    data = request.get_json()

    # Sanitize inputs
    username = sanitize_input(data.get("username", ""))
    email = sanitize_input(data.get("email", ""))
    password = data.get("password", "")

    # Validate inputs
    if not username or not email or not password:
        return jsonify({"status": "error", "message": "Invalid input"}), 400

    if not validate_password(password):
        return jsonify({"status": "error", "message": "Password does not meet complexity requirements"}), 400

    # Hash password securely
    hashed_pw = bcrypt.generate_password_hash(password).decode("utf-8")

    try:
        conn = sqlite3.connect("users.db")
        cursor = conn.cursor()

        # Use parameterized query to prevent SQL injection
        cursor.execute(
            "INSERT INTO users (username, email, password) VALUES (?, ?, ?)",
            (username, email, hashed_pw)
        )
        conn.commit()
        conn.close()

        return jsonify({"status": "success", "message": "Registration successful"}), 201

    except sqlite3.IntegrityError:
        # Generic error message (do not reveal if username/email exists)
        return jsonify({"status": "error", "message": "Registration failed"}), 400

    except Exception:
        return jsonify({"status": "error", "message": "An unexpected error occurred"}), 500


if __name__ == "__main__":
    app.run(debug=True)
